tools/ossf_best_practices/passing_criteria.md

```diff
@@ -0,0 +1,397 @@
+
+**Met**
+
+
+> It is SUGGESTED that the project implement continuous integration (where new or changed code is frequently integrated into a central code repository and automated tests are run on the result).
+
+**Met**
+https://ci.nodejs.org/
+
+
+## New functionality testing
+
+> The project MUST have a general policy (formal or not) that as major new functionality is added to the software produced by the project, tests of that functionality should be added to an automated test suite.
+
+**Met** 
+https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-5-test
+
+> The project MUST have evidence that the test_policy for adding tests has been adhered to in the most recent major changes to the software produced by the project.
+
+**Met**
+
+> It is SUGGESTED that this policy on adding tests (see test_policy) be documented in the instructions for change proposals
+
+**Met**
+
+## Warning flags
+
+> The project MUST enable one or more compiler warning flags, a "safe" language mode, or use a separate "linter" tool to look for code quality errors or common simple mistakes, if there is at least one FLOSS tool that can implement this criterion in the selected language.
+
+**Met**
+-Wall for compiling, c++ and JS linting as part of test suite
+
+
+> The project MUST address warnings.
+
+**Met**
+
+> It is SUGGESTED that projects be maximally strict with warnings in the software produced by the project, where practical.
+
+**Met**
+
+# Security
+
+## Secure development knowledge
+
+> The project MUST have at least one primary developer who knows how to design secure software. (See ‘details’ for the exact requirements.)
+
+**Met**
+
+> At least one of the project's primary developers MUST know of common kinds of errors that lead to vulnerabilities in this kind of software, as well as at least one method to counter or mitigate each of them.
+
+**Met**
+
+## Use basic good cryptographic practices
+
+
+> The software produced by the project MUST use, by default, only cryptographic protocols and algorithms that are publicly published and reviewed by experts (if cryptographic protocols and algorithms are used).
+
+**Met**
+all crypto uses openssl
+
+> If the software produced by the project is an application or library, and its primary purpose is not to implement cryptography, then it SHOULD only call on software specifically designed to implement cryptographic functions; it SHOULD NOT re-implement its own.
+
+**Met**
+all crypto uses openssl
+
+> All functionality in the software produced by the project that depends on cryptography MUST be implementable using FLOSS.
+
+**Met**
+
+> The security mechanisms within the software produced by the project MUST use default keylengths that at least meet the NIST minimum requirements through the year 2030 (as stated in 2012). It MUST be possible to configure the software so that smaller keylengths are completely disabled.
+
+**Met**
```

Comment:
@tniessen would also be interested in your view on the crypto related questions that follow.

Reply:
Without going into much detail, Node.js 20 meets all of these criteria in some way.

The only questionable bit is this:

> It MUST be possible to configure the software so that smaller keylengths are completely disabled

For TLS, this is absolutely possible. But if we also count the crypto module as `security mechanisms within the software`, configuring Node.js such that application code cannot somehow instantiate weak cryptographic primitives isn't as simple. Small DH (not ECDH) group are a notorious example. Node.js does not use them itself, but we also cannot easily prevent users from using them. Then again, if someone wants to, they can use custom OpenSSL configurations, custom OpenSSL providers, or even custom dynamically linked OpenSSL builds to "configure" Node.js's crypto module.

Reply:
@tniessen Even with FIPS enabled?

Reply:
@richardlau If I remember correctly (and that's a big if, I might be entirely wrong about this), the OpenSSL 3 FIPS module only makes guarantees for applications that don't use any deprecated or low-level APIs. We still use a variety of deprecated and low-level APIs, and we support legacy engines.

Reply:
FYI I'm currently poking at FIPS and Node.js with an aim of adding some testing into the CI. The reason I asked is that I'm seeing a lot of test failures with FIPS enabled, mostly complaining about key sizes.

Reply:
I think we can have a good discussion about it in the next wg meeting. I will move the discussion to https://github.com/nodejs/security-wg/issues/988